jupyterlab/tests/test_labextensions.py

```diff
@@ -0,0 +1,321 @@
+                d.cleanup()
+
+        self.src = self.tempdir()
+        self.name = 'mockextension'
+        self.files = files = [
+            pjoin(u'ƒile'),
+            pjoin(u'∂ir', u'ƒile1'),
+            pjoin(u'∂ir', u'∂ir2', u'ƒile2'),
+        ]
+        for file in files:
+            fullpath = os.path.join(self.src, file)
+            parent = os.path.dirname(fullpath)
+            if not os.path.exists(parent):
+                os.makedirs(parent)
+            touch(fullpath)
+
+        self.test_dir = self.tempdir()
+        self.data_dir = os.path.join(self.test_dir, 'data')
+        self.config_dir = os.path.join(self.test_dir, 'config')
+        self.system_data_dir = os.path.join(self.test_dir, 'system_data')
+        self.system_path = [self.system_data_dir]
+        self.system_labext = os.path.join(self.system_data_dir, 'labextensions')
+
+        # Patch out os.environ so that tests are isolated from the real OS
+        # environment.
+        self.patch_env = patch.dict('os.environ', {
+            'JUPYTER_CONFIG_DIR': self.config_dir,
+            'JUPYTER_DATA_DIR': self.data_dir,
+        })
+        self.patch_env.start()
+        self.addCleanup(self.patch_env.stop)
+
+        # Patch out the system path os that we consistently use our own
+        # temporary directory instead.
+        self.patch_system_path = patch.object(
+            labextensions, 'SYSTEM_JUPYTER_PATH', self.system_path
+        )
+        self.patch_system_path.start()
+        self.addCleanup(self.patch_system_path.stop)
+
+    def assert_dir_exists(self, path):
+        if not os.path.exists(path):
+            do_exist = os.listdir(os.path.dirname(path))
+            self.fail(u"%s should exist (found %s)" % (path, do_exist))
+    
+    def assert_not_dir_exists(self, path):
+        if os.path.exists(path):
+            self.fail(u"%s should not exist" % path)
+    
+    def assert_installed(self, relative_path, user=False):
+        if user:
+            labext = pjoin(self.data_dir, u'labextensions')
+        else:
+            labext = self.system_labext
+        self.assert_dir_exists(
+            pjoin(labext, relative_path)
+        )
+    
+    def assert_not_installed(self, relative_path, user=False):
+        if user:
+            labext = pjoin(self.data_dir, u'labextensions')
+        else:
+            labext = self.system_labext
+        self.assert_not_dir_exists(
+            pjoin(labext, relative_path)
+        )
+    
+    def test_create_labextensions_user(self):
+        with TemporaryDirectory() as td:
+            install_labextension(self.src, self.name, user=True)
+            self.assert_installed(
+                pjoin(self.name),
+                user=True
+            )
+    
+    def test_create_labextensions_system(self):
+        with TemporaryDirectory() as td:
+            self.system_labext = pjoin(td, u'labextensions')
+            with patch.object(labextensions, 'SYSTEM_JUPYTER_PATH', [td]):
```

Comment:
Not this particular directory, I tried removing it and it claimed it did not exist.